### ISSUE TITLE
Mark positioning

### ISSUE THREAD

**Issue by be5invis** [rsms/inter#155]
It would be great the Inter supports IPA and other scripts / notation systems that relies on combining marks.
![image](https://user-images.githubusercontent.com/240091/57438081-b1214b00-7275-11e9-91c0-35474a4a0e45.png)

**Comment by be5invis** [comment]
IPA is a great example of using combining diacritics:
![image](https://user-images.githubusercontent.com/240091/62437099-ad775600-b6f6-11e9-98fa-bdb2b93c2763.png)

**Comment by be5invis** [comment]
@schriftgestalt Glyphs should already support automatically export `mark` and `mkmk` lookups?
I found this article → https://glyphsapp.com/tutorials/mark-attachment

**Comment by rsms** [comment]
It appears as glyphs only creates mark and mkmk tables when using the GUI export feature — Inter uses it’s own separate build system based on fontmake and glyphslib, so we don’t get this for free.

Further, when I export from the Glyphs app using its GUI, the OTF files are noticeably larger than the OTF files produces by Inter’s fontmake-based build system. Additionally, Glyphs presents the following error on export (but it succeeds in creating fully-validating OTF files) “In feature 'mark' lookup 'marklatn' mark to base positioning rules cause an offset overflow (0x10a68) to a lookup subtable”.

So, based on this, I think we need to manually create a mark table. Or, patch glyphslib to generate it for us. Either way, it seems to be a lot of work.

Perhaps worth it? What would be the primary upsides of this? Is IPA commonly used and would Inter be able to have a significant impact on IPA use?

**Comment by schriftgestalt** [comment]
As fare as I know, glyphsLib+fontMake should support automatic mark+mkmk feature generation.

**Comment by be5invis** [comment]
> “In feature 'mark' lookup 'marklatn' mark to base positioning rules cause an offset overflow (0x10a68) to a lookup subtable”

@schriftgestalt  Hmm I think you can try to bisect these mark positioning lookups when they grow too large (like bisect at size > 32K).

**Comment by schriftgestalt** [comment]
mark lookups do not support subtables. 
Adding a "Use Extension Kerning" custom parameter in Font Info > Font should fix the problem.

**Comment by be5invis** [comment]
@schriftgestalt 
> mark lookups do not support subtables.
Source? I don't see the GPOS spec disallowed multiple-subtable mark lookup.

**Comment by be5invis** [comment]
@schriftgestalt At least SIL’s Arabic font *Scheherazade Regular* contains a mark-to-base lookup with two subtables.

**Comment by schriftgestalt** [comment]
You are right. But I don't think that a subtable break would help here. The mark feature itself is not the problem but the GPOS in total. Mostly the kerning is filling the 64k space and then the mark feature is just the last straw.

**Comment by anthrotype** [comment]
> As fare as I know, glyphsLib+fontMake should support automatic mark+mkmk feature generation.

Correct. If you're having any issues let us know!

**Comment by KrasnayaPloshchad** [comment]
> As fare as I know, glyphsLib+fontMake should support automatic mark+mkmk feature generation.

Wikipedia suggests the `ccmp` feature can also be used (see [here](https://en.wikipedia.org/wiki/Combining_character#OpenType)). BTW what about AFDKO?

**Comment by rsms** [comment]
Made some progress today. Here’s what Safari now renders:

<img width="462" alt="Screen Shot 2019-09-07 at 14 54 04" src="https://user-images.githubusercontent.com/47046/64480602-93121b80-d17f-11e9-8fc8-415509ca70f7.png">

This is the input:

```d U+030C U+0304 U+0323 U+0326
y U+030D U+0311 U+0310 U+0302 U+0301 U+0353 U+0347
n U+0306 U+0308 U+1DD8 U+0304 U+0307 U+032D U+0332 U+032B U+0323
a U+0363 U+0306 U+0309 U+0324 U+0330 U+032A
m U+0310 U+0318
i U+0304 U+1DCA
c U+0319 U+030B

d U+0325 U+0309
i U+030E U+1DC7
a U+1DC8 U+031E
c U+0348 U+0332
r U+0346 U+0332
i U+0313 U+036F
t U+032B U+0306
i U+1DC5 U+0307
c U+0368 U+0301
s U+031C U+1DCC```

**Comment by rsms** [comment]
> > As fare as I know, glyphsLib+fontMake should support automatic mark+mkmk feature generation.
> 
> Wikipedia suggests the `ccmp` feature can also be used (see [here](https://en.wikipedia.org/wiki/Combining_character#OpenType)). BTW what about AFDKO?

[ccmp](https://docs.microsoft.com/en-us/typography/opentype/spec/features_ae#ccmp) is used already: [src/features/ccmp.fea](https://github.com/rsms/inter/blob/49cd0c56c80c87ed721a3a317e891348f4515da1/src/features/ccmp.fea) :–)

**Comment by rsms** [comment]
There’s still a _ton of work_ left to be done, adding input and output anchors to every combining mark. (Only some currently have anchors)

**Comment by schriftgestalt** [comment]
You get pretty good default anchors when you select all glyphs and press Cmd+U.

**Comment by KrasnayaPloshchad** [comment]
I have seen an example of double stacked marks with the same diacritics, for example, [this article](https://en.wikipedia.org/wiki/File:JianningBible1.gif) has an example of double stacked macron in an image, have you thought of adding such implementation?

**Comment by KrasnayaPloshchad** [comment]
I have seen an example of double stacked marks with the same diacritics, for example, [this article](https://en.wikipedia.org/wiki/File:JianningBible1.gif) has an example of double stacked macron in an image, have you thought of adding such implementation?

**Comment by KrasnayaPloshchad** [comment]
The combining dot above right shouldn‘t break mark placing with other diacritics. You can see [this article](https://en.wikipedia.org/wiki/O%CD%98) for more details.

**Comment by KrasnayaPloshchad** [comment]
Combining vertical line above should having more mark attachments to be aligned with base letter, and having small spacing with them.
![图片](https://user-images.githubusercontent.com/10364351/109538116-abf05880-7afa-11eb-9808-03405b9b8d74.png)
Combining diaerasis below failed to aligned with base letter.
![图片](https://user-images.githubusercontent.com/10364351/109536889-3b950780-7af9-11eb-8365-8a891d40e6e2.png)
AE-monograph need more mark attachments
![图片](https://user-images.githubusercontent.com/10364351/109537581-ff15db80-7af9-11eb-8025-6fd82b568ec7.png)

See:
https://en.wikipedia.org/wiki/Pha%CC%8Dk-fa-s%E1%B9%B3
https://en.wikipedia.org/wiki/Foochow_Romanized
https://en.wikipedia.org/wiki/Hainan_Romanized
https://en.wikipedia.org/wiki/Hinghwa_Romanized

**Comment by VladWinner** [comment]
Can something similar be implemented to align the diacritics of Cyrillic vowels?
![image](https://user-images.githubusercontent.com/13555921/124192696-5d65c180-dace-11eb-99dd-5c407d02daf3.png)
Sample text:
```
Latin:   áéóý
Cyrillic: а́е́о́у́
А́а́, Е́е́, И́и́, О́о́, У́у́, Ы́ы́, Э́э́, Ю́ю́, Я́я́
```

**Comment by KrasnayaPloshchad** [comment]
I found a variant of ỉ on the _Dictionarium Annamiticum Lusitanum et Latinum_ (i.e. “Dictionary of Vietnamese, Portuguese and Latin”) that ratains the tittle, in this case the hook is placed at the right of character frame, and the altitude of hook is sightly lower. To implement this variant, you can just modify the mark attachment for 0069 0307 0309.

See: https://commons.wikimedia.org/wiki/File:Alexandre_de_Rhodes_-_Dictionarium_Annamiticum_(1651)_-_c.34_-_cropped_on_b%C3%AD_and_b%E1%BB%89.png

**Comment by bluebear94** [comment]
I’d also like for digits to be supported, so I can use things like `N4̂te`.

![image](https://user-images.githubusercontent.com/2975203/151720093-aa64b4b4-4019-4573-b5d3-daa765621ffb.png)

**Comment by brainandforce** [comment]
Double acute does not render correctly over A a (A̋ a̋), even though double grave does.
